src/awkward/_v2/forms/numpyform.py

```diff
@@ -0,0 +1,147 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+from __future__ import absolute_import
+
+try:
+    from collections.abc import Iterable
+except ImportError:
+    from collections import Iterable
+
+import copy
+
+from awkward._v2.forms.form import Form
+
+import awkward as ak
+
+np = ak.nplike.NumpyMetadata.instance()
+
+_primitive_to_dtype = {
+    "bool": np.dtype(np.bool_),
+    "int8": np.dtype(np.int8),
+    "uint8": np.dtype(np.uint8),
+    "int16": np.dtype(np.int16),
+    "uint16": np.dtype(np.uint16),
+    "int32": np.dtype(np.int32),
+    "uint32": np.dtype(np.uint32),
+    "int64": np.dtype(np.int64),
+    "uint64": np.dtype(np.uint64),
+    "float32": np.dtype(np.float32),
+    "float64": np.dtype(np.float64),
+    "complex64": np.dtype(np.complex64),
+    "complex128": np.dtype(np.complex128),
+    "datetime64": np.dtype(np.datetime64),
+    "timedelta64": np.dtype(np.timedelta64),
+}
+
+if hasattr(np, "float16"):
+    _primitive_to_dtype["float16"] = np.dtype(np.float16)
+if hasattr(np, "float128"):
+    _primitive_to_dtype["float128"] = np.dtype(np.float128)
+if hasattr(np, "complex256"):
+    _primitive_to_dtype["complex256"] = np.dtype(np.complex256)
+
+_dtype_to_primitive = {}
+for primitive, dtype in _primitive_to_dtype.items():
+    _dtype_to_primitive[dtype] = primitive
+
+
+def from_dtype(dtype, parameters={}, inner_shape=[]):
+    if "[" in str(dtype):
+        dtype = str(dtype).split("[")
+        # WIP
+        params = copy.deepcopy(parameters)
+        params["__unit__"] = dtype[1][:-1]
+        dtype = dtype[0]
```

Comment:
I get that this is a work in progress, but the units are only going to exist in some `np.datetime64` and `np.timedelta64` (not all), so you can narrow to that for safety. (I don't know if other dtypes might use square brackets for other things.)

The defensive copy of `parameters` can be shallow, and dict has a handy `.copy()` method.

The `inner_shape` also comes from the dtype, or it can. (I don't remember whether I added examples of that to the tests.) That's if the dtype's `subdtype` is not `None`.

```python
>>> dt = np.dtype((np.int64, (1, 2, 3)))
>>> dt
dtype(('<i8', (1, 2, 3)))
>>> dt.shape
(1, 2, 3)
>>> dt.subdtype
(dtype('int64'), (1, 2, 3))
```

This function should probably check `_dtype_to_primitive` to make sure that the dtype is one of the allowed dtypes in the list (after breaking it down with `subdtype`, if need be). Here are some dtypes that we should not accept (with a good error message):

```python
>>> np.dtype("O")
dtype('O')
>>> np.dtype([("one", np.int64), ("two", np.float64)])
dtype([('one', '<i8'), ('two', '<f8')])
```

Reply:
Unsupported dtypes are caught by __init__, I have added some tests to check if a TypeError is raised on unsupported types

Reply:
This `from_dtype` function is not in the NumpyForm class; it's not necessarily called after any `__init__`.